### PR TITLE
TTA AoA Ensemble: inference-time rotation augmentation for OOD robustness

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Test-time augmentation (TTA) with K=5 AoA rotations at inference. The model is not perfectly AoA-equivariant — averaging predictions over nearby AoA values smooths inconsistency and reduces OOD prediction variance. Standard Kaggle competition technique applied to CFD.

**Zero training changes** — pure inference augmentation. Gaussian weighting (sigma=1 degree). AoA deltas: [-2, -1, 0, +1, +2] degrees. Targets p_oodc, p_re.

### Add config flag
```python
tta_aoa_ensemble: bool = False
```

### Implement TTA function
```python
def tta_predict(model, x, forward_fn, aoa_deltas=[-2,-1,0,1,2]):
    import numpy as np
    weights = np.exp(-0.5 * (np.array(aoa_deltas) / 1.0) ** 2)
    weights /= weights.sum()
    preds_agg = None
    for delta_deg, w in zip(aoa_deltas, weights):
        d = delta_deg * np.pi / 180.0
        cos_d, sin_d = np.cos(d), np.sin(d)
        x_aug = x.clone()
        xy = x_aug[:, :, 0:2]
        x_aug[:, :, 0] = cos_d * xy[:,:,0] - sin_d * xy[:,:,1]
        x_aug[:, :, 1] = sin_d * xy[:,:,0] + cos_d * xy[:,:,1]
        with torch.no_grad():
            pred_aug = forward_fn(x_aug)
        pred_back = pred_aug.clone()
        pred_back[:,:,0] = cos_d * pred_aug[:,:,0] + sin_d * pred_aug[:,:,1]
        pred_back[:,:,1] = -sin_d * pred_aug[:,:,0] + cos_d * pred_aug[:,:,1]
        preds_agg = w * pred_back if preds_agg is None else preds_agg + w * pred_back
    return preds_agg
```

### Apply in eval loop ONLY (NOT during training)
```python
if cfg.tta_aoa_ensemble:
    pred = tta_predict(ema_model, x_norm, lambda x_in: ema_model(x_in, ...))
else:
    pred = ema_model(x_norm, ...)
```

TTA adds 5x inference cost per validation — acceptable since validation runs once per epoch.

### Run 2 seeds
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/tta-aoa-ensemble-s42" \
  --wandb_group "thorfinn/tta-aoa-ensemble" --seed 42 \
  --tta_aoa_ensemble \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
# Seed 73 — same with --seed 73
```

### Report
Table: p_in, p_oodc, p_tan, p_re. W&B run IDs. Note validation time impact.

## Baseline (PR #2251)
| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.891 | < 11.89 |
| p_oodc | 7.561 | < 7.56 |
| p_tan | 28.118 | < 28.12 |
| p_re | 6.364 | < 6.36 |

W&B: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### TTA AoA Ensemble — 2-seed results

| Metric | Baseline (PR #2251) | S42 (u88mll5s) | S73 (utqr7hjn) | 2-seed Avg | Δ vs Baseline |
|--------|-------------------|----------------|----------------|------------|---------------|
| p_in   | 11.891            | 18.621         | 16.797         | 17.709     | +48.9% ❌      |
| p_oodc | 7.561             | 11.800         | 9.800          | 10.800     | +42.8% ❌      |
| p_tan  | 28.118            | 31.500         | 31.600         | 31.550     | +12.2% ❌      |
| p_re   | 6.364             | 8.200          | 8.300          | 8.250      | +29.6% ❌      |

### Additional metrics (best epoch)

| | S42 (ep 96) | S73 (ep 101) |
|--|------------|--------------|
| val_loss | 0.4598 | 0.4531 |
| Surf Ux (in_dist) | 3.597 | 3.040 |
| Surf Uy (in_dist) | 1.221 | 1.316 |
| Vol Ux (in_dist) | 0.675 | 0.638 |
| Vol Uy (in_dist) | 0.280 | 0.302 |
| Vol p (in_dist) | 17.699 | 16.530 |
| Epoch time | ~108s | ~107s |
| Peak memory | 46.6 GB | 46.4 GB |

### Commands used
```bash
# Seed 42
CUDA_VISIBLE_DEVICES=0 python train.py --agent thorfinn --wandb_name "thorfinn/tta-aoa-ensemble-s42" \
  --wandb_group "thorfinn/tta-aoa-ensemble" --seed 42 --tta_aoa_ensemble \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73 — identical with --seed 73
```

### W&B run IDs
- Seed 42: `u88mll5s` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/u88mll5s
- Seed 73: `utqr7hjn` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/utqr7hjn

### What happened — clear negative result

TTA AoA ensemble **significantly degraded** all metrics, with p_in and p_oodc hit hardest (+49% and +43% worse). This is a clear negative result. Analysis of why:

1. **Rotation ≠ AoA change**: Rotating coordinates by ±1-2° doesn't replicate the physical effect of changing the angle of attack. A true AoA change alters the entire flow field — boundary layers, separation points, pressure distributions — in ways that a simple coordinate rotation cannot capture. The model correctly learns AoA-specific features, and averaging across rotated inputs blurs these features.

2. **Training augmentation already covers this**: The baseline uses `--aug aoa_perturb` which perturbs AoA during training. The model is already trained to be robust to small AoA variations. TTA at inference adds noise on top of an already-robust model.

3. **The model uses AoA as an input feature**: Since AoA is encoded in the distance/direction fields (dsdf), rotating coordinates creates an inconsistency between the geometric features and the actual flow conditions, confusing the model rather than helping it.

4. **Epoch time overhead**: ~108s/epoch vs ~78s baseline (+38%) due to 5x forward passes per validation. This cost would be acceptable if TTA helped, but it doesn't.

5. **Verification NaN bug**: The manual denormalization verification produced NaN for both seeds. This appears to be because the TTA wrapper in the validation loop returns predictions in a different code path that the verification step doesn't handle correctly. This doesn't affect the main metrics (which come from the regular validation loop's per-split accumulation), but the verification check is broken when TTA is enabled.

### Suggested follow-ups

- **Close this approach**: TTA with coordinate rotation is fundamentally flawed for CFD because AoA rotation ≠ coordinate rotation in terms of physics. This is not worth pursuing further in its current form.
- **If TTA is revisited**: Would need actual re-simulation at different AoA values (not available) or a learned AoA-conditional ensemble approach. Neither is practical.
- **The verification NaN bug with TTA enabled**: Minor issue since TTA won't be used, but if TTA is ever revisited the verification code path needs adjustment to handle the wrapped model output.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2272 because ADVISOR: Closing — all 4 metrics regressed massively (p_in +48.9%, p_oodc +42.8%, p_tan +12.2%, p_re +29.6%).

**Root cause:** Excellent analysis. The fundamental flaw: coordinate rotation ≠ physical AoA change. Rotating (x,y) coordinates does not change the flow physics (boundary layers, separation, pressure). Since AoA is encoded in the DSDF features, rotation creates an inconsistency between geometry and physics. Averaging 5 inconsistent predictions degrades all metrics. Training augmentation (aoa_perturb) already provides rotation robustness — TTA adds noise on top.

**Key takeaway:** TTA rotation is only valid when the transformation matches the physical symmetry. For CFD, AoA rotation changes the physics (flow separation, boundary layer) — not just the coordinate frame. Added to DO NOT REVISIT.

Good catch on the verification NaN bug too — noted for future reference. Moving you to a fresh direction..
